### PR TITLE
Pwm patch + AnalogOut updates

### PR DESCRIPTION
**Pull request by jamesadevine**
Fixes for #115 #218.

Enables AnalogOut on all digital capable IO.

Pushes functionality from the previous DynamicPWM driver into the mbed pwm driver, reducing the size of the DynamicPwm class.

Here is a sample test program to test the swapping of PWMs and also a primitive test of the new capabilities:
```cpp
#include "MicroBit.h"

MicroBit uBit;

MicroBitPin* pins[4] = {&uBit.io.P0, &uBit.io.P1, &uBit.io.P2, &uBit.io.P8};

int main()
{
    // Initialise the micro:bit runtime.
    uBit.init();

    int i = 0;

    while(1)
    {
        pins[i]->setAnalogValue(512);

        uBit.display.print(i);
        uBit.sleep(5000);

        i = (i + 1) % 4;
    }

    // Insert your code here!
    uBit.display.scroll("HELLO WORLD! :)");

    // If main exits, there may still be other fibers running or registered event handlers etc.
    // Simply release this fiber, which will mean we enter the scheduler. Worse case, we then
    // sit in the idle task forever, in a power efficient sleep.
    release_fiber();
}
```

Would be great if I could get a second pair of eyes on this, it can be tested by tagging against mbed-classic#master and microbit-dal#pwm-patch

### CODE REVIEW

File: source/drivers/DynamicPwm.cpp [hunk 1]
```diff
@@ -89,82 +43,9 @@ void gpiote_reinit(PinName pin, PinName oldPin, uint8_t channel_number)
...
-    if(pwms[lastUsed]->flags & PWM_PERSISTENCE_TRANSIENT)
-    {
-        pwms[lastUsed]->flags = persistence;
-        pwms[lastUsed]->redirect(pin);
-        return pwms[lastUsed];
-    }
-
-    //well if we have no transient channels - we can't give any away! :( return null
-    return (DynamicPwm*)NULL;
+    this->period = 0;
```
Comment (finneyj) [inline source/drivers/DynamicPwm.cpp]: Do we really want 0 as the default? I seem to recall we default to 20ms at the moment? (servo frequency)?

**Comment by jamesadevine** [comment]
I think a default of zero is better than presuming that the period is actually 20 ms, it may not be. The developer can check for zero, then ensure that the period is their desired frequency.


James.

On 2 Feb 2017, at 09:34, Joe Finney <notifications@github.com<mailto:notifications@github.com>> wrote:


@finneyj commented on this pull request.

________________________________

In source/drivers/DynamicPwm.cpp<https://github.com/lancaster-university/microbit-dal/pull/264#pullrequestreview-19742268>:

> -        }
-
-        channelIterator = (channelIterator + 1 > NO_PWMS - 1) ? 0 : channelIterator + 1;
-    }
-
-    //if we haven't found a free one, we must try to allocate the last used...
-    if(pwms[lastUsed]->flags & PWM_PERSISTENCE_TRANSIENT)
-    {
-        pwms[lastUsed]->flags = persistence;
-        pwms[lastUsed]->redirect(pin);
-        return pwms[lastUsed];
-    }
-
-    //well if we have no transient channels - we can't give any away! :( return null
-    return (DynamicPwm*)NULL;
+    this->period = 0;


Do we really want 0 as the default? I seem to recall we default to 20ms at the moment? (servo frequency)?

**Comment by jamesadevine** [comment]
If this were static and shared across all instances of DynamicPwm, it would make sense to default to 20 ms


James.

On 2 Feb 2017, at 09:34, Joe Finney <notifications@github.com<mailto:notifications@github.com>> wrote:


@finneyj commented on this pull request.

________________________________

In source/drivers/DynamicPwm.cpp<https://github.com/lancaster-university/microbit-dal/pull/264#pullrequestreview-19742268>:

> -        }
-
-        channelIterator = (channelIterator + 1 > NO_PWMS - 1) ? 0 : channelIterator + 1;
-    }
-
-    //if we haven't found a free one, we must try to allocate the last used...
-    if(pwms[lastUsed]->flags & PWM_PERSISTENCE_TRANSIENT)
-    {
-        pwms[lastUsed]->flags = persistence;
-        pwms[lastUsed]->redirect(pin);
-        return pwms[lastUsed];
-    }
-
-    //well if we have no transient channels - we can't give any away! :( return null
-    return (DynamicPwm*)NULL;
+    this->period = 0;


Do we really want 0 as the default? I seem to recall we default to 20ms at the moment? (servo frequency)?

**Comment by finneyj** [comment]
yes, that's what i was thinking... 0 as a period is never useful as a default.

**Comment by jamesadevine** [comment]
Although, we are just duplicating state here... 

Accessors should be exposed by mbed for this sort of thing, we could add one...?

**Comment by finneyj** [comment]
You're quite right, but let's try not to diverge the mbed fork any more than we need to...

**Comment by DaveAtKitronik** [comment]
test code for this issue:
`#include "MicroBit.h"

MicroBit uBit;

// we use events abd the 'connected' variable to keep track of the status of the Bluetooth connection
void onConnected(MicroBitEvent)
{
    uBit.display.print("C");
}

void onDisconnected(MicroBitEvent)
{
    uBit.display.print("D");
}


void onButton(MicroBitEvent e)
{
    if ((e.source == MICROBIT_ID_BUTTON_A)&&(e.value == MICROBIT_BUTTON_EVT_DOWN))
	{
        uBit.display.print("A");
		uBit.io.P0.setAnalogValue(500); //set the volume to about 50%
		uBit.io.P0.setAnalogPeriodUs(3846); // ~260Hz
	}
 
    if ((e.source == MICROBIT_ID_BUTTON_B)&&(e.value == MICROBIT_BUTTON_EVT_DOWN))
	{
		uBit.display.print("B");
		uBit.io.P0.setAnalogValue(500);
		uBit.io.P0.setAnalogPeriodUs(3413); //~293 hz
	}

    if ((e.source == MICROBIT_ID_BUTTON_AB)&&(e.value == MICROBIT_BUTTON_EVT_DOWN))
	{
		uBit.io.P0.setAnalogValue(0);
		uBit.display.scroll("A+B");
    }
}


int main()
{
    // Initialise the micro:bit runtime.
    uBit.init();

    // Configuration Tips
    //
    // config.json contains various Bluetooth related properties some of which are explained here:
    //
    // "dfu_service": 1,               // 1 will cause the DFU service to be instantiated
    // "event_service": 1,             // 1 causes the event service to be instantiated
    // "device_info_service": 1
    // "enabled": 1,                   // 1 means the Bluetooth stack will be included as standard. 0 means it will not.
    // "pairing_mode": 1,              // 1 means it's possible to go into pairing mode which will include bringing up the Bluetooth stack whilst in that mode.
    // "open": 0,                      // 1 means there's no Bluetooth security i.e. no need to pair the micro:bit with other devices that want to communicate with it.
    // "tx_power": 7,                  // Transmission power of the Bluetooth radio. A value of 0 - 7 with 0 the lowest power and 7 the highest power.
    // "gatt_table_size": "0x700"      // Amount of memory (in hex bytes) set aside for the Bluetooth GATT table
    // "nested_heap_proportion": 0.75, // Reducing this can sometimes help make enough memory available for all the Bluetooth services you want. Only experiment with this as a last resort.

    // MicrobitConfig.h in yotta_modules\microbit-dal\inc\core contains MICROBIT_BLE_SECURITY_LEVEL which can be set to SECURITY_MODE_ENCRYPTION_WITH_MITM for passkey authentication when
    // pairing or SECURITY_MODE_ENCRYPTION_NO_MITM to use Just Works pairing.

    // A cunning code to indicate during start-up the particular Bluetooth configuration in the build
    //
    // SERVICE CODES
    // A: Accelerometer Service
    // B: Button Service
    // D: Device Information Service
    // E: Event Service
    // F: DFU Service
    // I: IO Pin Service
    // L: LED Service
    // M: Magnetometer Service
    // T: Temperature Service
    // U: UART Service
    //
    // PAIRING CONFIG
    // Note that switching pairing on or off is achieved by setting "open" in config.json to 1 or 0 respectively

    // P: PASSKEY
    // J: Just Works
    // N: No Pairing Required
    //
    // TX Power Level
    // 0-7 taken from tx_power in config.json


    // Services/Pairing Config/Power Level
    uBit.display.scroll("BLE ABDILMT/P/0");

    uBit.messageBus.listen(MICROBIT_ID_BLE, MICROBIT_BLE_EVT_CONNECTED, onConnected);
    uBit.messageBus.listen(MICROBIT_ID_BLE, MICROBIT_BLE_EVT_DISCONNECTED, onDisconnected);
	
	uBit.messageBus.listen(MICROBIT_ID_BUTTON_A, MICROBIT_EVT_ANY, onButton);
    uBit.messageBus.listen(MICROBIT_ID_BUTTON_B, MICROBIT_EVT_ANY, onButton);
	uBit.messageBus.listen(MICROBIT_ID_BUTTON_AB, MICROBIT_EVT_ANY, onButton);

    new MicroBitAccelerometerService(*uBit.ble, uBit.accelerometer);
    //new MicroBitButtonService(*uBit.ble);
    new MicroBitIOPinService(*uBit.ble, uBit.io);
    new MicroBitLEDService(*uBit.ble, uBit.display);
    new MicroBitMagnetometerService(*uBit.ble, uBit.compass);
    new MicroBitTemperatureService(*uBit.ble, uBit.thermometer);

    // If main exits, there may still be other fibers running or registered event handlers etc.
    // Simply release this fiber, which will mean we enter the scheduler. Worse case, we then
    // sit in the idle task forever, in a power efficient sleep.
    release_fiber();
}`
and config.json in the root of the build:
`{
    "microbit-dal": {
        "bluetooth": {
            "enabled": 1,
            "pairing_mode": 1,
            "private_addressing": 0,
            "open": 1,
            "whitelist": 1,
            "advertising_timeout": 0,
            "tx_power": 7,
            "dfu_service": 0,
            "event_service": 0,
            "device_info_service": 1
        },
        "gatt_table_size": "0x700"
    }
}
`
This generates a couple of tones that showed the problem in blocks (approx. 100uS jitter).
PXT has ~12Us jitter.
Have yet to measure the jitter on this code

**Comment by DaveAtKitronik** [comment]
Could not get the Bluetooth to run correctly with the exisiting code, so cut down to just accel service
```
int main()
{
    // Initialise the micro:bit runtime.
    uBit.init();

    // Configuration Tips
    //
    // config.json contains various Bluetooth related properties some of which are explained here:
    //
    // "dfu_service": 1,               // 1 will cause the DFU service to be instantiated
    // "event_service": 1,             // 1 causes the event service to be instantiated
    // "device_info_service": 1
    // "enabled": 1,                   // 1 means the Bluetooth stack will be included as standard. 0 means it will not.
    // "pairing_mode": 1,              // 1 means it's possible to go into pairing mode which will include bringing up the Bluetooth stack whilst in that mode.
    // "open": 0,                      // 1 means there's no Bluetooth security i.e. no need to pair the micro:bit with other devices that want to communicate with it.
    // "tx_power": 7,                  // Transmission power of the Bluetooth radio. A value of 0 - 7 with 0 the lowest power and 7 the highest power.
    // "gatt_table_size": "0x700"      // Amount of memory (in hex bytes) set aside for the Bluetooth GATT table
    // "nested_heap_proportion": 0.75, // Reducing this can sometimes help make enough memory available for all the Bluetooth services you want. Only experiment with this as a last resort.

    // MicrobitConfig.h in yotta_modules\microbit-dal\inc\core contains MICROBIT_BLE_SECURITY_LEVEL which can be set to SECURITY_MODE_ENCRYPTION_WITH_MITM for passkey authentication when
    // pairing or SECURITY_MODE_ENCRYPTION_NO_MITM to use Just Works pairing.

    // A cunning code to indicate during start-up the particular Bluetooth configuration in the build
    //
    // SERVICE CODES
    // A: Accelerometer Service
    // B: Button Service
    // D: Device Information Service
    // E: Event Service
    // F: DFU Service
    // I: IO Pin Service
    // L: LED Service
    // M: Magnetometer Service
    // T: Temperature Service
    // U: UART Service
    //
    // PAIRING CONFIG
    // Note that switching pairing on or off is achieved by setting "open" in config.json to 1 or 0 respectively

    // P: PASSKEY
    // J: Just Works
    // N: No Pairing Required
    //
    // TX Power Level
    // 0-7 taken from tx_power in config.json


    // Services/Pairing Config/Power Level
    uBit.display.scroll("BLE A/P/0");

    uBit.messageBus.listen(MICROBIT_ID_BLE, MICROBIT_BLE_EVT_CONNECTED, onConnected);
    uBit.messageBus.listen(MICROBIT_ID_BLE, MICROBIT_BLE_EVT_DISCONNECTED, onDisconnected);
	
	uBit.messageBus.listen(MICROBIT_ID_BUTTON_A, MICROBIT_EVT_ANY, onButton);
    uBit.messageBus.listen(MICROBIT_ID_BUTTON_B, MICROBIT_EVT_ANY, onButton);
	uBit.messageBus.listen(MICROBIT_ID_BUTTON_AB, MICROBIT_EVT_ANY, onButton);

    new MicroBitAccelerometerService(*uBit.ble, uBit.accelerometer);

    // If main exits, there may still be other fibers running or registered event handlers etc.
    // Simply release this fiber, which will mean we enter the scheduler. Worse case, we then
    // sit in the idle task forever, in a power efficient sleep.
    release_fiber();
}
```
This will successfully connect to the micro:bit Blue app on a Samsung S4, and reports the accelerometer data as expected.
The tone generation is as before, and shows no jitter, either by listening, or on a scope. 
Only tested on 260 and 293 hz, but I see no reason that different frequencies here this would affect the initial problem.
Fix is good for C++. now need to get into the blocks and PXT world I guess

**Comment by jamesadevine** [comment]
Awesome testing thanks @DaveAtKitronik :smile:

**Comment by jamesadevine** [comment]
@finneyj static member variable added.

**Comment by jamesadevine** [comment]
Code to verify:

```cpp
#include "MicroBit.h"

MicroBit uBit;

MicroBitPin* pins[4] = {&uBit.io.P0, &uBit.io.P1, &uBit.io.P2, &uBit.io.P8};

int main()
{
    // Initialise the micro:bit runtime.
    uBit.init();
    pins[0]->setAnalogValue(90);
    pins[1]->setAnalogValue(90);
    uBit.display.scroll(pins[0]->getAnalogPeriodUs());
    pins[0]->setAnalogPeriodUs(10000);
    uBit.display.scroll(pins[1]->getAnalogPeriodUs());
    int i = 0;

    while(1)
    {
        pins[i]->setAnalogValue(512);

        uBit.display.print(i);
        uBit.sleep(5000);

        i = (i + 1) % 4;
    }
}

```

**Pull request merged**